File: scenarios/advanced/desktop/macos-swiftui/n5.swift
import SwiftUI

struct SocialMediaFeed: View {
    @State private var posts: [Post] = []
    @State private var allPosts: [Post] = []
    @State private var currentFilter = "All"
    @State private var unreadNotifications = 3
    @State private var createPostText = ""
    @State private var showingCreatePost = false
    @State private var showingNotifications = false
    @State private var showingTrending = false
    @State private var selectedPost: Post?
    
    var body: some View {
        VStack(spacing: 0) {
            // Header
            headerView
            
            // Create Post
            createPostView
            
            // Posts Feed
            postsFeedView
            
            // Status Bar
            statusBarView
        }
        .onAppear {
            loadSamplePosts()
        }
    }
    
    private var headerView: some View {
        HStack {
            Text("Social Media Feed")
                .font(.largeTitle)
                .fontWeight(.bold)
                .foregroundColor(.white)
            
            Spacer()
            
            HStack(spacing: 12) {
                Picker("Filter", selection: $currentFilter) {
                    Text("All").tag("All")
                    Text("Following").tag("Following")
                    Text("Trending").tag("Trending")
                }
                .pickerStyle(.menu)
                .frame(width: 140)
                
                Button("🔔") {
                    showingNotifications = true
                }
                .buttonStyle(.bordered)
                .overlay(
                    Circle()
                        .fill(Color.red)
                        .frame(width: 12, height: 12)
                        .offset(x: 8, y: -8)
                        .opacity(unreadNotifications > 0 ? 1 : 0)
                )
                
                Button("📈") {
                    showingTrending = true
                }
                .buttonStyle(.bordered)
            }
        }
        .padding(20)
        .background(Color.accentColor)
    }
    
    private var createPostView: some View {
        HStack(spacing: 16) {
            // Avatar
            Circle()
                .fill(Color.accentColor)
                .frame(width: 48, height: 48)
                .overlay(
                    Text("👤")
                        .font(.title2)
                )
            
            // Text Field
            HStack {
                TextField("What's on your mind?", text: $createPostText, axis: .vertical)
                    .textFieldStyle(.plain)
                    .lineLimit(1...4)
                
                Button("Post") {
                    createPost()
                }
                .buttonStyle(.borderedProminent)
                .disabled(createPostText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .background(Color(.controlBackgroundColor))
            .cornerRadius(8)
        }
        .padding(.horizontal, 24)
        .padding(.vertical, 16)
        .background(Color(.controlBackgroundColor))
    }
    
    private var postsFeedView: some View {
        ScrollView {
            LazyVStack(spacing: 16) {
                ForEach(posts, id: \.id) { post in
                    PostCard(post: post) { action in
                        handlePostAction(action, for: post)
                    }
                }
            }
            .padding(.horizontal, 24)
        }
    }
    
    private var statusBarView: some View {
        HStack {
            Text("Ready")
            
            Spacer()
            
            Text("\(posts.count) posts")
                .font(.caption)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        .background(Color(.controlBackgroundColor))
    }
    
    // MARK: - Methods
    private func loadSamplePosts() {
        allPosts = [
            Post(
                id: "1",
                user: User(id: "user1", name: "John Doe", username: "@johndoe", avatar: "👨‍💻", verified: true),
                content: "Just finished building an amazing React Native app! The development process was challenging but incredibly rewarding. #ReactNative #MobileDev #Tech",
                image: "📱",
                location: "San Francisco, CA",
                timestamp: "2h ago",
                likes: 42,
                comments: 8,
                shares: 3,
                isLiked: false,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "2",
                user: User(id: "user2", name: "Sarah Wilson", username: "@sarahw", avatar: "👩‍🎨", verified: false),
                content: "Beautiful sunset from my office window today. Sometimes you need to stop and appreciate the little moments. 🌅",
                image: "🌅",
                location: "New York, NY",
                timestamp: "4h ago",
                likes: 28,
                comments: 5,
                shares: 1,
                isLiked: true,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "3",
                user: User(id: "user3", name: "Mike Chen", username: "@mikechen", avatar: "👨‍💼", verified: true),
                content: "Excited to announce our new product launch! After months of hard work, we're finally ready to share it with the world. #Startup #Innovation",
                image: "🚀",
                location: "Austin, TX",
                timestamp: "6h ago",
                likes: 156,
                comments: 23,
                shares: 12,
                isLiked: false,
                isBookmarked: true,
                privacy: "public"
            ),
            Post(
                id: "4",
                user: User(id: "user4", name: "Emma Davis", username: "@emmad", avatar: "👩‍🔬", verified: false),
                content: "Just read an amazing article about the future of AI in healthcare. The possibilities are endless! What are your thoughts on this?",
                image: "🤖",
                location: "Boston, MA",
                timestamp: "8h ago",
                likes: 73,
                comments: 15,
                shares: 7,
                isLiked: true,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "5",
                user: User(id: "user5", name: "Alex Rodriguez", username: "@alexr", avatar: "👨‍🎓", verified: false),
                content: "Coffee and code - the perfect combination for a productive day! ☕️💻",
                image: "☕️",
                location: "Seattle, WA",
                timestamp: "10h ago",
                likes: 34,
                comments: 6,
                shares: 2,
                isLiked: false,
                isBookmarked: false,
                privacy: "public"
            )
        ]
        
        posts = allPosts
    }
    
    private func createPost() {
        let content = createPostText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return }
        
        let newPost = Post(
            id: UUID().uuidString,
            user: User(id: "current-user", name: "You", username: "@you", avatar: "👤", verified: false),
            content: content,
            image: "",
            location: "",
            timestamp: "now",
            likes: 0,
            comments: 0,
            shares: 0,
            isLiked: false,
            isBookmarked: false,
            privacy: "public"
        )
        
        posts.insert(newPost, at: 0)
        allPosts.insert(newPost, at: 0)
        createPostText = ""
    }
    
    private func handlePostAction(_ action: PostAction, for post: Post) {
        switch action {
        case .like:
            toggleLike(for: post)
        case .comment:
            showComments(for: post)
        case .share:
            sharePost(post)
        case .bookmark:
            toggleBookmark(for: post)
        }
    }
    
    private func toggleLike(for post: Post) {
        if let index = posts.firstIndex(where: { $0.id == post.id }) {
            posts[index].isLiked.toggle()
            posts[index].likes += posts[index].isLiked ? 1 : -1
            
            // Update in allPosts as well
            if let allPostsIndex = allPosts.firstIndex(where: { $0.id == post.id }) {
                allPosts[allPostsIndex].isLiked = posts[index].isLiked
                allPosts[allPostsIndex].likes = posts[index].likes
            }
        }
    }
    
    private func showComments(for post: Post) {
        selectedPost = post
        // In a real implementation, this would show a comments view
    }
    
    private func sharePost(_ post: Post) {
        // Share implementation
    }
    
    private func toggleBookmark(for post: Post) {
        if let index = posts.firstIndex(where: { $0.id == post.id }) {
            posts[index].isBookmarked.toggle()
            
            // Update in allPosts as well
            if let allPostsIndex = allPosts.firstIndex(where: { $0.id == post.id }) {
                allPosts[allPostsIndex].isBookmarked = posts[index].isBookmarked
            }
        }
    }
}

// MARK: - Supporting Views
struct PostCard: View {
    let post: Post
    let onAction: (PostAction) -> Void
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Header
            HStack(spacing: 12) {
                // Avatar
                Text(post.user.avatar)
                    .font(.title)
                
                // User Info
                VStack(alignment: .leading, spacing: 2) {
                    HStack(spacing: 4) {
                        Text(post.user.name)
                            .font(.headline)
                            .fontWeight(.bold)
                        
                        if post.user.verified {
                            Image(systemName: "checkmark.circle.fill")
                                .foregroundColor(.blue)
                                .font(.caption)
                        }
                    }
                    
                    Text(post.user.username)
                        .font(.subheadline)
                        .foregroundColor(.secondary)
                    
                    Text(post.timestamp)
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
                
                Spacer()
            }
            
            // Content
            Text(post.content)
                .font(.body)
            
            // Image
            if !post.image.isEmpty {
                Text(post.image)
                    .font(.system(size: 48))
                    .frame(maxWidth: .infinity)
                    .padding()
                    .background(Color(.controlBackgroundColor))
                    .cornerRadius(8)
            }
            
            // Location
            if !post.location.isEmpty {
                HStack(spacing: 4) {
                    Image(systemName: "location")
                        .foregroundColor(.secondary)
                        .font(.caption)
                    
                    Text(post.location)
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }
            
            // Actions
            HStack(spacing: 16) {
                ActionButton(
                    icon: post.isLiked ? "heart.fill" : "heart",
                    text: "\(post.likes)",
                    color: post.isLiked ? .red : .secondary
                ) {
                    onAction(.like)
                }
                
                ActionButton(
                    icon: "message",
                    text: "\(post.comments)",
                    color: .secondary
                ) {
                    onAction(.comment)
                }
                
                ActionButton(
                    icon: "arrow.clockwise",
                    text: "\(post.shares)",
                    color: .secondary
                ) {
                    onAction(.share)
                }
                
                ActionButton(
                    icon: post.isBookmarked ? "bookmark.fill" : "bookmark",
                    text: "",
                    color: post.isBookmarked ? .blue : .secondary
                ) {
                    onAction(.bookmark)
                }
                
                Spacer()
            }
        }
        .padding(16)
        .background(Color(.controlBackgroundColor))
        .cornerRadius(12)
    }
}

struct ActionButton: View {
    let icon: String
    let text: String
    let color: Color
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            HStack(spacing: 4) {
                Image(systemName: icon)
                    .font(.system(size: 16))
                
                if !text.isEmpty {
                    Text(text)
                        .font(.caption)
                }
            }
            .foregroundColor(color)
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Notifications View
struct NotificationsView: View {
    @Binding var isPresented: Bool
    let unreadCount: Int
    
    var body: some View {
        VStack(spacing: 20) {
            Text("Notifications")
                .font(.title)
                .fontWeight(.bold)
            
            Text("You have \(unreadCount) unread notifications")
                .font(.body)
                .foregroundColor(.secondary)
            
            // In a real implementation, this would show actual notifications
            VStack(alignment: .leading, spacing: 8) {
                NotificationRow(title: "New like on your post", time: "2m ago")
                NotificationRow(title: "John Doe commented on your post", time: "5m ago")
                NotificationRow(title: "Sarah Wilson shared your post", time: "1h ago")
            }
            
            HStack {
                Button("Mark All Read") {
                    // Mark all as read
                }
                .buttonStyle(.bordered)
                
                Button("Close") {
                    isPresented = false
                }
                .buttonStyle(.borderedProminent)
            }
        }
        .padding(20)
        .frame(width: 400, height: 300)
    }
}

struct NotificationRow: View {
    let title: String
    let time: String
    
    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: 4) {
                Text(title)
                    .font(.body)
                
                Text(time)
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
            
            Spacer()
        }
        .padding(8)
        .background(Color(.controlBackgroundColor))
        .cornerRadius(6)
    }
}

// MARK: - Trending View
struct TrendingView: View {
    @Binding var isPresented: Bool
    
    var body: some View {
        VStack(spacing: 20) {
            Text("Trending")
                .font(.title)
                .fontWeight(.bold)
            
            VStack(alignment: .leading, spacing: 12) {
                TrendingTopic(topic: "#ReactNative", posts: "2.3K posts")
                TrendingTopic(topic: "#MobileDev", posts: "1.8K posts")
                TrendingTopic(topic: "#Tech", posts: "3.1K posts")
                TrendingTopic(topic: "#Startup", posts: "1.2K posts")
                TrendingTopic(topic: "#Innovation", posts: "956 posts")
            }
            
            HStack {
                Button("Close") {
                    isPresented = false
                }
                .buttonStyle(.borderedProminent)
                
                Spacer()
            }
        }
        .padding(20)
        .frame(width: 300, height: 250)
    }
}

struct TrendingTopic: View {
    let topic: String
    let posts: String
    
    var body: some View {
        HStack {
            Text(topic)
                .font(.headline)
                .foregroundColor(.accentColor)
            
            Spacer()
            
            Text(posts)
                .font(.caption)
                .foregroundColor(.secondary)
        }
        .padding(8)
        .background(Color(.controlBackgroundColor))
        .cornerRadius(6)
    }
}

// MARK: - Models
struct User {
    let id: String
    let name: String
    let username: String
    let avatar: String
    let verified: Bool
}

struct Post: Identifiable {
    let id: String
    let user: User
    let content: String
    let image: String
    let location: String
    let timestamp: String
    var likes: Int
    let comments: Int
    let shares: Int
    var isLiked: Bool
    var isBookmarked: Bool
    let privacy: String
    
    var title: String {
        return content.count > 50 ? String(content.prefix(50)) + "..." : content
    }
}

enum PostAction {
    case like
    case comment
    case share
    case bookmark
}

// MARK: - Preview
struct SocialMediaFeed_Previews: PreviewProvider {
    static var previews: some View {
        SocialMediaFeed()
    }
}

File: scenarios/advanced/desktop/linux-qt/n5.py
#!/usr/bin/env python3

import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QWidget, QVBoxLayout, 
                             QHBoxLayout, QGridLayout, QLabel, QPushButton, 
                             QComboBox, QListWidget, QListWidgetItem, QLineEdit,
                             QCheckBox, QMessageBox, QFrame, QScrollArea,
                             QGroupBox, QSlider, QSpinBox, QTextEdit, QDialog)
from PyQt5.QtCore import Qt, QTimer, pyqtSignal, QRect
from PyQt5.QtGui import QFont, QPalette, QColor, QIcon, QPainter, QPen, QBrush
from datetime import datetime

class SocialMediaFeed(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Social Media Feed")
        self.setGeometry(100, 100, 1000, 700)
        
        # Data
        self.posts = []
        self.all_posts = []
        self.current_filter = "All"
        self.unread_notifications = 3
        
        # Load sample data
        self.load_sample_posts()
        
        # Create UI
        self.create_ui()
    
    def create_ui(self):
        # Central widget
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        # Main layout
        main_layout = QVBoxLayout()
        central_widget.setLayout(main_layout)
        
        # Header
        header_widget = self.create_header()
        main_layout.addWidget(header_widget)
        
        # Create post
        create_post_widget = self.create_post_box()
        main_layout.addWidget(create_post_widget)
        
        # Posts feed
        self.create_posts_feed(main_layout)
        
        # Status bar
        self.status_bar = self.statusBar()
        self.status_bar.showMessage("Ready")
    
    def create_header(self):
        header_widget = QFrame()
        header_widget.setStyleSheet("""
            QFrame {
                background-color: #4A90E2;
                padding: 20px;
                border: none;
            }
        """)
        
        header_layout = QHBoxLayout()
        header_widget.setLayout(header_layout)
        
        # Title
        title_label = QLabel("Social Media Feed")
        title_label.setStyleSheet("""
            QLabel {
                color: white;
                font-size: 24px;
                font-weight: bold;
            }
        """)
        header_layout.addWidget(title_label)
        
        header_layout.addStretch()
        
        # Controls
        controls_layout = QHBoxLayout()
        
        # Filter combo
        self.filter_combo = QComboBox()
        self.filter_combo.addItems(["All", "Following", "Trending"])
        self.filter_combo.currentTextChanged.connect(self.on_filter_changed)
        controls_layout.addWidget(self.filter_combo)
        
        # Notifications button
        self.notifications_btn = QPushButton("🔔")
        self.notifications_btn.setToolTip("Notifications")
        self.notifications_btn.setStyleSheet("""
            QPushButton {
                background-color: rgba(255, 255, 255, 0.2);
                color: white;
                border: none;
                padding: 8px 12px;
                border-radius: 4px;
                font-size: 16px;
            }
            QPushButton:hover {
                background-color: rgba(255, 255, 255, 0.3);
            }
        """)
        self.notifications_btn.clicked.connect(self.on_notifications_clicked)
        controls_layout.addWidget(self.notifications_btn)
        
        # Trending button
        trending_btn = QPushButton("📈")
        trending_btn.setToolTip("Trending")
        trending_btn.setStyleSheet("""
            QPushButton {
                background-color: rgba(255, 255, 255, 0.2);
                color: white;
                border: none;
                padding: 8px 12px;
                border-radius: 4px;
                font-size: 16px;
            }
            QPushButton:hover {
                background-color: rgba(255, 255, 255, 0.3);
            }
        """)
        trending_btn.clicked.connect(self.on_trending_clicked)
        controls_layout.addWidget(trending_btn)
        
        header_layout.addLayout(controls_layout)
        
        return header_widget
    
    def create_post_box(self):
        create_post_widget = QFrame()
        create_post_widget.setStyleSheet("""
            QFrame {
                background-color: #F8F8F8;
                padding: 16px;
                border: none;
            }
        """)
        
        create_post_layout = QHBoxLayout()
        create_post_widget.setLayout(create_post_layout)
        
        # Avatar
        avatar_label = QLabel("👤")
        avatar_label.setStyleSheet("font-size: 32px;")
        create_post_layout.addWidget(avatar_label)
        
        # Text entry and button
        entry_layout = QHBoxLayout()
        
        self.create_post_entry = QLineEdit()
        self.create_post_entry.setPlaceholderText("What's on your mind?")
        self.create_post_entry.setStyleSheet("""
            QLineEdit {
                padding: 12px;
                border: 2px solid #E0E0E0;
                border-radius: 8px;
                font-size: 14px;
                background-color: white;
            }
            QLineEdit:focus {
                border-color: #4A90E2;
            }
        """)
        self.create_post_entry.returnPressed.connect(self.on_create_post)
        entry_layout.addWidget(self.create_post_entry)
        
        self.create_post_btn = QPushButton("Post")
        self.create_post_btn.setStyleSheet("""
            QPushButton {
                background-color: #4A90E2;
                color: white;
                border: none;
                padding: 12px 24px;
                border-radius: 8px;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #357ABD;
            }
            QPushButton:disabled {
                background-color: #CCCCCC;
            }
        """)
        self.create_post_btn.clicked.connect(self.on_create_post)
        entry_layout.addWidget(self.create_post_btn)
        
        create_post_layout.addLayout(entry_layout)
        
        return create_post_widget
    
    def create_posts_feed(self, parent):
        # Scroll area for posts
        scroll_area = QScrollArea()
        scroll_area.setWidgetResizable(True)
        scroll_area.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        scroll_area.setHorizontalScrollBarPolicy(Qt.ScrollBarNever)
        
        # Posts container
        self.posts_widget = QWidget()
        self.posts_layout = QVBoxLayout()
        self.posts_widget.setLayout(self.posts_layout)
        
        scroll_area.setWidget(self.posts_widget)
        parent.addWidget(scroll_area)
        
        # Update posts display
        self.update_posts_display()
    
    def load_sample_posts(self):
        self.all_posts = [
            {
                'id': '1',
                'user': {
                    'id': 'user1',
                    'name': 'John Doe',
                    'username': '@johndoe',
                    'avatar': '👨‍💻',
                    'verified': True
                },
                'content': 'Just finished building an amazing React Native app! The development process was challenging but incredibly rewarding. #ReactNative #MobileDev #Tech',
                'image': '📱',
                'location': 'San Francisco, CA',
                'timestamp': '2h ago',
                'likes': 42,
                'comments': 8,
                'shares': 3,
                'is_liked': False,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '2',
                'user': {
                    'id': 'user2',
                    'name': 'Sarah Wilson',
                    'username': '@sarahw',
                    'avatar': '👩‍🎨',
                    'verified': False
                },
                'content': 'Beautiful sunset from my office window today. Sometimes you need to stop and appreciate the little moments. 🌅',
                'image': '🌅',
                'location': 'New York, NY',
                'timestamp': '4h ago',
                'likes': 28,
                'comments': 5,
                'shares': 1,
                'is_liked': True,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '3',
                'user': {
                    'id': 'user3',
                    'name': 'Mike Chen',
                    'username': '@mikechen',
                    'avatar': '👨‍💼',
                    'verified': True
                },
                'content': 'Excited to announce our new product launch! After months of hard work, we\'re finally ready to share it with the world. #Startup #Innovation',
                'image': '🚀',
                'location': 'Austin, TX',
                'timestamp': '6h ago',
                'likes': 156,
                'comments': 23,
                'shares': 12,
                'is_liked': False,
                'is_bookmarked': True,
                'privacy': 'public'
            },
            {
                'id': '4',
                'user': {
                    'id': 'user4',
                    'name': 'Emma Davis',
                    'username': '@emmad',
                    'avatar': '👩‍🔬',
                    'verified': False
                },
                'content': 'Just read an amazing article about the future of AI in healthcare. The possibilities are endless! What are your thoughts on this?',
                'image': '🤖',
                'location': 'Boston, MA',
                'timestamp': '8h ago',
                'likes': 73,
                'comments': 15,
                'shares': 7,
                'is_liked': True,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '5',
                'user': {
                    'id': 'user5',
                    'name': 'Alex Rodriguez',
                    'username': '@alexr',
                    'avatar': '👨‍🎓',
                    'verified': False
                },
                'content': 'Coffee and code - the perfect combination for a productive day! ☕️💻',
                'image': '☕️',
                'location': 'Seattle, WA',
                'timestamp': '10h ago',
                'likes': 34,
                'comments': 6,
                'shares': 2,
                'is_liked': False,
                'is_bookmarked': False,
                'privacy': 'public'
            }
        ]
        
        self.posts = self.all_posts.copy()
    
    def update_posts_display(self):
        # Clear existing posts
        for i in reversed(range(self.posts_layout.count())):
            self.posts_layout.itemAt(i).widget().setParent(None)
        
        # Add posts
        for post in self.posts:
            post_widget = self.create_post_widget(post)
            self.posts_layout.addWidget(post_widget)
        
        # Update status bar
        self.status_bar.showMessage(f"{len(self.posts)} posts")
    
    def create_post_widget(self, post):
        # Main container
        post_frame = QFrame()
        post_frame.setStyleSheet("""
            QFrame {
                background-color: white;
                border: 1px solid #E0E0E0;
                border-radius: 8px;
                margin: 8px;
            }
        """)
        
        post_layout = QVBoxLayout()
        post_frame.setLayout(post_layout)
        
        # Header
        header_layout = QHBoxLayout()
        
        # Avatar
        avatar_label = QLabel(post['user']['avatar'])
        avatar_label.setStyleSheet("font-size: 32px;")
        header_layout.addWidget(avatar_label)
        
        # User info
        user_info_layout = QVBoxLayout()
        
        # Name and verification
        name_layout = QHBoxLayout()
        
        name_label = QLabel(post['user']['name'])
        name_label.setFont(QFont("Arial", 14, QFont.Bold))
        name_layout.addWidget(name_label)
        
        if post['user']['verified']:
            verified_label = QLabel("✓")
            verified_label.setStyleSheet("color: #4A90E2; font-weight: bold;")
            name_layout.addWidget(verified_label)
        
        name_layout.addStretch()
        
        user_info_layout.addLayout(name_layout)
        
        # Username and timestamp
        username_label = QLabel(f"{post['user']['username']} • {post['timestamp']}")
        username_label.setStyleSheet("color: #666666; font-size: 12px;")
        user_info_layout.addWidget(username_label)
        
        header_layout.addLayout(user_info_layout)
        header_layout.addStretch()
        
        post_layout.addLayout(header_layout)
        
        # Content
        content_label = QLabel(post['content'])
        content_label.setWordWrap(True)
        content_label.setStyleSheet("font-size: 14px; padding: 8px 0;")
        post_layout.addWidget(content_label)
        
        # Image
        if post['image']:
            image_label = QLabel(post['image'])
            image_label.setAlignment(Qt.AlignCenter)
            image_label.setStyleSheet("font-size: 48px; padding: 16px; background-color: #F8F8F8; border-radius: 8px;")
            post_layout.addWidget(image_label)
        
        # Location
        if post['location']:
            location_layout = QHBoxLayout()
            
            location_icon = QLabel("📍")
            location_layout.addWidget(location_icon)
            
            location_label = QLabel(post['location'])
            location_label.setStyleSheet("color: #666666; font-size: 12px;")
            location_layout.addWidget(location_label)
            
            location_layout.addStretch()
            
            post_layout.addLayout(location_layout)
        
        # Actions
        actions_layout = QHBoxLayout()
        
        # Like button
        like_btn = QPushButton()
        like_icon = "❤️" if post['is_liked'] else "🤍"
        like_btn.setText(f"{like_icon} {post['likes']}")
        like_btn.setStyleSheet("""
            QPushButton {
                background-color: transparent;
                border: none;
                padding: 8px 12px;
                text-align: left;
            }
            QPushButton:hover {
                background-color: #F0F0F0;
                border-radius: 4px;
            }
        """)
        like_btn.clicked.connect(lambda: self.on_like_clicked(post))
        actions_layout.addWidget(like_btn)
        
        # Comment button
        comment_btn = QPushButton(f"💬 {post['comments']}")
        comment_btn.setStyleSheet("""
            QPushButton {
                background-color: transparent;
                border: none;
                padding: 8px 12px;
                text-align: left;
            }
            QPushButton:hover {
                background-color: #F0F0F0;
                border-radius: 4px;
            }
        """)
        comment_btn.clicked.connect(lambda: self.on_comment_clicked(post))
        actions_layout.addWidget(comment_btn)
        
        # Share button
        share_btn = QPushButton(f"🔄 {post['shares']}")
        share_btn.setStyleSheet("""
            QPushButton {
                background-color: transparent;
                border: none;
                padding: 8px 12px;
                text-align: left;
            }
            QPushButton:hover {
                background-color: #F0F0F0;
                border-radius: 4px;
            }
        """)
        share_btn.clicked.connect(lambda: self.on_share_clicked(post))
        actions_layout.addWidget(share_btn)
        
        # Bookmark button
        bookmark_btn = QPushButton()
        bookmark_icon = "🔖" if post['is_bookmarked'] else "📌"
        bookmark_btn.setText(bookmark_icon)
        bookmark_btn.setStyleSheet("""
            QPushButton {
                background-color: transparent;
                border: none;
                padding: 8px 12px;
                text-align: left;
            }
            QPushButton:hover {
                background-color: #F0F0F0;
                border-radius: 4px;
            }
        """)
        bookmark_btn.clicked.connect(lambda: self.on_bookmark_clicked(post))
        actions_layout.addWidget(bookmark_btn)
        
        actions_layout.addStretch()
        
        post_layout.addLayout(actions_layout)
        
        return post_frame
    
    # Event handlers
    def on_filter_changed(self, filter_text):
        self.current_filter = filter_text
        # In a real implementation, this would filter posts
        self.status_bar.showMessage(f"Filter changed to {filter_text}")
    
    def on_notifications_clicked(self):
        dialog = NotificationsDialog(self, self.unread_notifications)
        dialog.exec_()
    
    def on_trending_clicked(self):
        dialog = TrendingDialog(self)
        dialog.exec_()
    
    def on_create_post(self):
        content = self.create_post_entry.text().strip()
        if not content:
            return
        
        new_post = {
            'id': str(datetime.now().timestamp()),
            'user': {
                'id': 'current-user',
                'name': 'You',
                'username': '@you',
                'avatar': '👤',
                'verified': False
            },
            'content': content,
            'image': '',
            'location': '',
            'timestamp': 'now',
            'likes': 0,
            'comments': 0,
            'shares': 0,
            'is_liked': False,
            'is_bookmarked': False,
            'privacy': 'public'
        }
        
        self.posts.insert(0, new_post)
        self.all_posts.insert(0, new_post)
        self.create_post_entry.clear()
        self.update_posts_display()
    
    def on_like_clicked(self, post):
        # Toggle like
        post['is_liked'] = not post['is_liked']
        post['likes'] += 1 if post['is_liked'] else -1
        
        # Update in all_posts as well
        for all_post in self.all_posts:
            if all_post['id'] == post['id']:
                all_post['is_liked'] = post['is_liked']
                all_post['likes'] = post['likes']
                break
        
        self.update_posts_display()
    
    def on_comment_clicked(self, post):
        QMessageBox.information(self, "Comments", f"Comments for: {post['user']['name']}\n\n{post['content']}")
    
    def on_share_clicked(self, post):
        QMessageBox.information(self, "Share", f"Sharing: {post['content'][:50]}...")
    
    def on_bookmark_clicked(self, post):
        # Toggle bookmark
        post['is_bookmarked'] = not post['is_bookmarked']
        
        # Update in all_posts as well
        for all_post in self.all_posts:
            if all_post['id'] == post['id']:
                all_post['is_bookmarked'] = post['is_bookmarked']
                break
        
        self.update_posts_display()

class NotificationsDialog(QDialog):
    def __init__(self, parent, unread_count):
        super().__init__(parent)
        self.setWindowTitle("Notifications")
        self.setModal(True)
        self.resize(400, 300)
        self.unread_count = unread_count
        
        self.setup_ui()
    
    def setup_ui(self):
        layout = QVBoxLayout()
        self.setLayout(layout)
        
        # Title
        title_label = QLabel("Notifications")
        title_label.setFont(QFont("Arial", 18, QFont.Bold))
        title_label.setAlignment(Qt.AlignCenter)
        layout.addWidget(title_label)
        
        # Unread count
        count_label = QLabel(f"You have {self.unread_count} unread notifications")
        count_label.setAlignment(Qt.AlignCenter)
        count_label.setStyleSheet("color: #666666;")
        layout.addWidget(count_label)
        
        # Notifications list
        notifications_list = QListWidget()
        notifications = [
            "New like on your post",
            "John Doe commented on your post",
            "Sarah Wilson shared your post"
        ]
        
        for notification in notifications:
            item = QListWidgetItem(notification)
            notifications_list.addItem(item)
        
        layout.addWidget(notifications_list)
        
        # Buttons
        button_layout = QHBoxLayout()
        
        mark_read_btn = QPushButton("Mark All Read")
        mark_read_btn.clicked.connect(self.mark_all_read)
        button_layout.addWidget(mark_read_btn)
        
        close_btn = QPushButton("Close")
        close_btn.clicked.connect(self.accept)
        button_layout.addWidget(close_btn)
        
        layout.addLayout(button_layout)
    
    def mark_all_read(self):
        QMessageBox.information(self, "Mark All Read", "All notifications marked as read.")
        self.accept()

class TrendingDialog(QDialog):
    def __init__(self, parent):
        super().__init__(parent)
        self.setWindowTitle("Trending")
        self.setModal(True)
        self.resize(300, 250)
        
        self.setup_ui()
    
    def setup_ui(self):
        layout = QVBoxLayout()
        self.setLayout(layout)
        
        # Title
        title_label = QLabel("Trending")
        title_label.setFont(QFont("Arial", 18, QFont.Bold))
        title_label.setAlignment(Qt.AlignCenter)
        layout.addWidget(title_label)
        
        # Trending topics
        topics_list = QListWidget()
        topics = [
            "#ReactNative - 2.3K posts",
            "#MobileDev - 1.8K posts",
            "#Tech - 3.1K posts",
            "#Startup - 1.2K posts",
            "#Innovation - 956 posts"
        ]
        
        for topic in topics:
            item = QListWidgetItem(topic)
            topics_list.addItem(item)
        
        layout.addWidget(topics_list)
        
        # Close button
        close_btn = QPushButton("Close")
        close_btn.clicked.connect(self.accept)
        layout.addWidget(close_btn)

def main():
    app = QApplication(sys.argv)
    
    # Set application style
    app.setStyle('Fusion')
    
    window = SocialMediaFeed()
    window.show()
    
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
